src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4954 @@
+			info.pi_space.ps_space.s_total[DAOS_MEDIA_NVME];
+	sfs->f_bfree = info.pi_space.ps_space.s_free[DAOS_MEDIA_SCM] +
+		       info.pi_space.ps_space.s_free[DAOS_MEDIA_NVME];
+	sfs->f_bsize  = 1;
+	sfs->f_files  = -1;
+	sfs->f_ffree  = -1;
+	sfs->f_bavail = sfs->f_bfree;
+
+	return 0;
+}
+
+int
+statfs64(const char *pathname, struct statfs64 *sfs) __attribute__((alias("statfs")));
+
+int
+__statfs(const char *pathname, struct statfs *sfs)
+	__attribute__((alias("statfs"), leaf, nonnull, nothrow));
+
+int
+statvfs(const char *pathname, struct statvfs *svfs)
+{
+	daos_pool_info_t info = {.pi_bits = DPI_SPACE};
+	dfs_obj_t       *parent;
+	int              rc, is_target_path;
+	struct dfs_mt   *dfs_mt;
+	char             item_name[DFS_MAX_NAME];
+	char             parent_dir[DFS_MAX_PATH];
+
+	if (next_statvfs == NULL) {
+		next_statvfs = dlsym(RTLD_NEXT, "statvfs");
+		D_ASSERT(next_statvfs != NULL);
+	}
+
+	if (!hook_enabled)
+		return next_statvfs(pathname, svfs);
+
+	rc = parse_path(pathname, &is_target_path, &parent, item_name, parent_dir, NULL, &dfs_mt);
+	if (rc)
+		return (-1);
+	if (!is_target_path)
+		return next_statvfs(pathname, svfs);
+
+	rc = daos_pool_query(dfs_mt->poh, NULL, &info, NULL, NULL);
+	if (rc) {
+		if (bLog)
+			D_ERROR("Failed to query pool: "DF_RC"\n", DP_RC(rc));
+		errno = daos_der2errno(rc);
+		return (-1);
+	}
+
+	svfs->f_blocks = info.pi_space.ps_space.s_total[DAOS_MEDIA_SCM] +
+			 info.pi_space.ps_space.s_total[DAOS_MEDIA_NVME];
+	svfs->f_bfree  = info.pi_space.ps_space.s_free[DAOS_MEDIA_SCM] +
+			 info.pi_space.ps_space.s_free[DAOS_MEDIA_NVME];
+	svfs->f_bsize  = 1;
+	svfs->f_files  = -1;
+	svfs->f_ffree  = -1;
+	svfs->f_bavail = svfs->f_bfree;
+
+	return 0;
+}
+
+int
+statvfs64(const char *__restrict pathname, struct statvfs64 *__restrict svfs)
+	__attribute__((alias("statvfs")));
+
+DIR *
+opendir(const char *path)
+{
+	int              is_target_path, idx_dirfd, rc;
+	dfs_obj_t       *parent, *dir_obj;
+	mode_t           mode;
+	struct dfs_mt   *dfs_mt;
+	char             item_name[DFS_MAX_NAME];
+	char             parent_dir[DFS_MAX_PATH];
+	char             full_path[DFS_MAX_PATH];
+
+	if (next_opendir == NULL) {
+		next_opendir = dlsym(RTLD_NEXT, "opendir");
+		D_ASSERT(next_opendir != NULL);
+	}
+	if (!hook_enabled)
+		return next_opendir(path);
+
+	rc = parse_path(path, &is_target_path, &parent, item_name, parent_dir, full_path, &dfs_mt);
+	if (rc)
+		return NULL;
+	if (!is_target_path)
+		return next_opendir(path);
+	if (bLog)
+		atomic_fetch_add_relaxed(&num_opendir, 1);
+
+	if (!parent && (strncmp(item_name, "/", 2) == 0))
+		rc = dfs_lookup(dfs_mt->dfs, "/", O_RDONLY, &dir_obj, &mode, NULL);
+	else
+		rc = dfs_open(dfs_mt->dfs, parent, item_name, S_IFDIR, O_RDONLY, 0, 0, NULL,
+			      &dir_obj);
+	if (rc) {
+		errno = rc;
+		return NULL;
+	}
+	mode = dfs_obj_mode(dfs_mt->dfs, dir_obj);
+	if ((S_IRUSR & mode) == 0)
+		D_GOTO(out_err, rc = EACCES);
```

Comment:
I thought dfs_open is less expensive to call than dfs_lookup. dfs_open() is mainly used here and it does not return mode directly.

Reply:
I see. right. 
but you are not calling that on the root. just calling dfs_lookup on root.

so you don't need this new function you added since lookup gives you mode for root and you can use the old function on object from dfs_open().

Reply:
ok. I am going to fix it. Thank you!

Reply:
Thank you! Fixed as you suggested.